snuba/consumer.py

```diff
@@ -110,3 +134,209 @@ def flush_batch(self, batch: Sequence[ProcessedMessage]):
+        assert not self.__closed
+
+        self.__messages.append(message)
+
+    def __delivery_callback(self, error, message) -> None:
+        if error is not None:
+            # errors are KafkaError objects and inherit from BaseException
+            raise error
+
+    def close(self) -> None:
+        self.__closed = True
+
+        if not self.__messages:
+            return
+
+        for message in self.__messages:
+            batch = message.payload
+            key = batch.key.encode("utf-8")
+            for value in batch.values:
+                self.__producer.produce(
+                    self.__topic.name,
+                    key=key,
+                    value=rapidjson.dumps(value).encode("utf-8"),
+                    on_delivery=self.__delivery_callback,
+                )
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        args = []
+        if timeout is not None:
+            args.append(timeout)
+
+        self.__producer.flush(*args)
+
+
+class ProcessedMessageBatchWriter(
+    ProcessingStep[Union[None, InsertBatch, ReplacementBatch]]
+):
+    def __init__(
+        self,
+        insert_batch_writer: InsertBatchWriter,
+        replacement_batch_writer: Optional[ReplacementBatchWriter] = None,
+    ) -> None:
+        self.__insert_batch_writer = insert_batch_writer
+        self.__replacement_batch_writer = replacement_batch_writer
+
+        self.__closed = False
+
+    def poll(self) -> None:
+        self.__insert_batch_writer.poll()
+
+        if self.__replacement_batch_writer is not None:
+            self.__replacement_batch_writer.poll()
+
+    def submit(
+        self, message: Message[Union[None, InsertBatch, ReplacementBatch]]
+    ) -> None:
+        assert not self.__closed
+
+        if message.payload is None:
+            return
+
+        if isinstance(message.payload, InsertBatch):
+            self.__insert_batch_writer.submit(cast(Message[InsertBatch], message))
+        elif isinstance(message.payload, ReplacementBatch):
+            assert self.__replacement_batch_writer is not None
+            self.__replacement_batch_writer.submit(
+                cast(Message[ReplacementBatch], message)
+            )
+        else:
+            raise TypeError("unexpected payload type")
+
+    def close(self) -> None:
+        self.__closed = True
+
+        self.__insert_batch_writer.close()
+
+        if self.__replacement_batch_writer is not None:
+            self.__replacement_batch_writer.close()
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        start = time.time()
+        self.__insert_batch_writer.join(timeout)
+
+        if self.__replacement_batch_writer is not None:
+            if timeout is not None:
+                timeout = max(timeout - (time.time() - start), 0)
+
+            self.__replacement_batch_writer.join(timeout)
+
+
+class StreamingConsumerStrategyFactory(ProcessingStrategyFactory[KafkaPayload]):
```

Comment:
Ideally this would be generic on `TPayload` but that can't happen until `MessageProcessor` itself is generic.